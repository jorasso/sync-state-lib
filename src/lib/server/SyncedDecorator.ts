// Decorates the value of a property to be synced with the client
export const Synced = () => {
  return function (target: any, propertyKey: string) {
    // Initialize the static property if it doesn't exist
    if (!target.constructor.hasOwnProperty("SyncedProperties")) {
      target.constructor.SyncedProperties = []
    }

    if (!target.constructor.hasOwnProperty("AllSynced")) {
      target.constructor.AllSynced = []
    }

    // Add the property key to the list of synced properties
    target.constructor.SyncedProperties.push(propertyKey)
    target.constructor.AllSynced.push(propertyKey)
  }
}

// Decorates the value of a property to be a component synced with the client
export const SyncedComponent = () => {
  return function (target: any, propertyKey: string) {
    // Initialize the static property if it doesn't exist
    if (!target.constructor.hasOwnProperty("SyncedComponents")) {
      target.constructor.SyncedComponents = []
    }

    if (!target.constructor.hasOwnProperty("AllSynced")) {
      target.constructor.AllSynced = []
    }

    // Add the property key to the list of synced components
    target.constructor.SyncedComponents.push(propertyKey)
    target.constructor.AllSynced.push(propertyKey)
  }
}

// Decorates a class to be synced with the client.
// This is needed to avoid having to call makeSynced() manually.
// Otherwise defineProperty() doesn't work as expected with derived classes.
export const WithSynced = <T extends { new (...args: any[]): {} }>(
  constructor: T
) => {
  return class extends constructor {
    constructor(...args: any[]) {
      super(...args)
      //@ts-ignore
      if (typeof this["makeSynced"] === "function") {
        //@ts-ignore
        this["makeSynced"]()
      }
    }
  }
}
