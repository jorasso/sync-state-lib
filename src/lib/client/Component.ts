import { ComponentChangeType } from "../server/Component"
import { GetObjectProperty } from "../utils/GetObjectProperty"
import { SetObjectProperty } from "../utils/SetObjectProperty"

export class Component {
  static AllSyncedAlphabeticalOrder: string[] = []
  static AllSynced: string[] = []
  static SyncedComponents: string[] = []
  static SyncedProperties: string[] = []

  protected components = new Map<number, Component>()

  protected properties = new Map<number, string>()

  private listeners = new Map<string, Array<(value: any) => void>>()

  private parentComponent!: Component
  private parentNumId!: number
  private parentPropertyKey!: string

  makeSynced() {
    const properties: string[] =
      GetObjectProperty(this.constructor, "SyncedProperties") ?? []
    const components: string[] =
      GetObjectProperty(this.constructor, "SyncedComponents") ?? []
    const all: string[] = GetObjectProperty(this.constructor, "AllSynced") ?? []
    const allOrder: string[] = [...all].sort()

    all.forEach((propertyKey) => {
      const isProperty = properties.includes(propertyKey)
      const numId = allOrder.indexOf(propertyKey)

      if (isProperty) {
        this.properties.set(numId, propertyKey)
      } else {
        let component = GetObjectProperty(this, propertyKey) as Component

        this.components.set(numId, component)
      }
    })
  }

  setParent(
    parentComponent: Component,
    parentNumId: number,
    parentPropertyKey?: string
  ) {
    this.parentComponent = parentComponent
    this.parentNumId = parentNumId

    if (parentPropertyKey) {
      this.parentPropertyKey = parentPropertyKey
    }
  }

  getParentComponent() {
    return this.parentComponent
  }

  getParentNumId() {
    return this.parentNumId
  }

  getParentPropertyKey() {
    return this.parentPropertyKey
  }

  loadCompleteStateBKP(state: any[]) {
    if (state[0] !== ComponentChangeType.complete) {
      return
    }

    const valuesSize = state[1]

    for (let i = 1; i < state.length; i += 3) {
      const key = state[i]
      const numId = state[i + 1]
      const value = state[i + 2]

      const isComponent = i >= 1 + valuesSize * 3

      if (isComponent) {
        //@ts-ignore
        const component = this[key]
        this.components.set(numId, component)
        component.setParent(this, key, numId)

        component.loadCompleteState(value)
      } else {
        this.properties.set(numId, key)
        //@ts-ignore
        this[key] = value

        this.emitOn(key, value)
      }
    }
  }

  loadCompleteState(state: any[]) {
    if (state[0] !== ComponentChangeType.complete) {
      return
    }

    for (let i = 1; i < state.length; i += 2) {
      const numId = state[i]
      const data = state[i + 1]

      const isComponent = this.components.has(numId)

      if (isComponent) {
        const component = this.components.get(numId)
        component.setParent(this, numId)

        component.loadCompleteState(data)
      } else {
        const key = this.properties.get(numId)
        //@ts-ignore
        this[key] = data

        this.emitOn(key, data)
      }
    }
  }

  on(key: string, callback: (value: any) => void) {
    if (!this.listeners.has(key)) {
      this.listeners.set(key, [])
    }

    this.listeners.get(key).push(callback)
  }

  private emitOn = (key: string, value: any) => {
    if (!this.listeners.has(key)) {
      return
    }

    const listeners = this.listeners.get(key)

    for (const listener of listeners) {
      listener(value)
    }
  }

  applyRecentChanges(changes: any[]) {
    if (changes[0] === ComponentChangeType.complete) {
      this.loadCompleteState(changes)
      return
    }

    if (changes[0] !== ComponentChangeType.partial) {
      return
    }

    for (let i = 1; i < changes.length; i += 2) {
      const numId = changes[i]
      const data = changes[i + 1]

      const isComponent = this.components.has(numId)

      if (isComponent) {
        const component = this.components.get(numId)
        component.applyRecentChanges(data)
      } else if (this.properties.has(numId)) {
        const key = this.properties.get(numId)
        //@ts-ignore
        this[key] = data

        this.emitOn(key, data)
      }
    }
  }

  applyRecentChangesBKP(data: any[]) {
    if (data[0] === ComponentChangeType.complete) {
      this.loadCompleteState(data)
      return
    }

    if (data[0] !== ComponentChangeType.partial) {
      return
    }

    for (let i = 1; i < data.length; i += 2) {
      const numId = data[i]
      const value = data[i + 1]

      if (this.components.has(numId)) {
        const component = this.components.get(numId)
        component.applyRecentChanges(value)
      } else if (this.properties.has(numId)) {
        const key = this.properties.get(numId)
        const value = data[i + 1]

        SetObjectProperty(this, key, value)
        this.emitOn(key, value)
      }
    }
  }
}
