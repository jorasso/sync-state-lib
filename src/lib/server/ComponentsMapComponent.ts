import { Component, ComponentChangeType } from "./Component"

enum MapOperations {
  add = 0,
  remove = 1,
  update = 2,
  replace = 3,
}

export class ComponentsMapComponent<T extends Component> extends Component {
  private map = new Map<string, T>()

  // Just to have the same signature as client's ComponentsMapComponent
  constructor(_: { new (): T }) {
    super()
  }

  size() {
    return this.map.size
  }

  set(key: string, component: T) {
    const currentComponent = this.map.get(key)

    if (currentComponent === component) {
      return
    }

    this.addToComponents(key, component, currentComponent?.getParentNumId())

    const numId = component.getParentNumId()

    this.addChildComponentPropertyChange(numId, currentComponent, component, {
      key,
    })

    currentComponent?.dispose()

    this.map.set(key, component)
  }

  get(key: string) {
    return this.map.get(key)
  }

  has(key: string) {
    return this.map.has(key)
  }

  delete(key: string) {
    const currentComponent = this.map.get(key)

    this.addChildComponentPropertyChange(
      currentComponent.getParentNumId(),
      currentComponent,
      undefined
    )

    currentComponent?.dispose()
    this.map.delete(key)
    this.removeFromComponents(currentComponent)
  }

  addToComponents = (
    key: string,
    component: Component,
    existingNumId: number
  ) => {
    const numId = existingNumId ?? this.getNumId()
    this.components.set(component, numId)
    component.setParent(this, key, numId)
  }

  removeFromComponents = (component: Component) => {
    this.components.delete(component)
  }

  getCompleteState() {
    const state: any[] = [ComponentChangeType.complete]

    state.push(this.map.size)

    this.map.forEach((component, key) => {
      state.push(key)
      state.push(this.components.get(component))

      state.push(component.getCompleteState())
    })

    return state
  }

  getRecentChanges() {
    const state: any[] = [ComponentChangeType.partial]

    this.childComponentsChanges.forEach((change, numId) => {
      if (change.to === undefined && change.from !== undefined) {
        state.push(MapOperations.remove)
        state.push(numId)

        return
      }

      if (change.from === undefined && change.to !== undefined) {
        state.push(MapOperations.add)
        state.push(change.meta.key)
        state.push(numId)
        state.push(change.to.getCompleteState())
        return
      }

      if (change.from !== change.to) {
        state.push(MapOperations.replace)
        state.push(numId)
        state.push(change.to.getCompleteState())
        return
      }

      const recentChanges = change.to.getRecentChanges()

      if (recentChanges.length === 0) {
        return
      }

      state.push(MapOperations.update)
      state.push(numId)
      state.push(recentChanges)
    })

    return state
  }
}
