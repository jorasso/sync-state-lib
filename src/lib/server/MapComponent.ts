import { Component, ComponentChangeType } from "./Component"

type AllowedTypes = number | string | boolean

enum MapOperations {
  add = 0,
  remove = 1,
  update = 2,
}

export class MapComponent<T extends AllowedTypes> extends Component {
  private map = new Map<string, T>()

  set(key: string, value: T) {
    this.addToProperties(key)

    const numId = this.properties.get(key)

    this.addPrimitivePropertyChange(numId, this.map.get(key), value, { key })

    this.map.set(key, value)
  }

  get(key: string) {
    return this.map.get(key)
  }

  has(key: string) {
    return this.map.has(key)
  }

  delete(key: string) {
    const numId = this.properties.get(key)
    this.addPrimitivePropertyChange(numId, this.map.get(key), undefined)
    this.map.delete(key)
    this.removeFromProperties(key)
  }

  addToProperties = (key: string) => {
    if (this.map.has(key)) {
      return
    }

    const numId = this.getNumId()

    this.properties.set(key, numId)
  }

  removeFromProperties = (key: string) => {
    this.properties.delete(key)
  }

  size() {
    return this.map.size
  }

  getCompleteState() {
    const state: any = [ComponentChangeType.complete]

    state.push(this.map.size)

    this.map.forEach((value, key) => {
      state.push(key)
      state.push(this.properties.get(key))

      state.push(value)
    })

    return state
  }

  getRecentChanges() {
    const state: any[] = [ComponentChangeType.partial]

    this.primitivePropertiesChanges.forEach((change, numId) => {
      if (change.from === change.to) {
        return
      }

      if (change.to === undefined) {
        state.push(MapOperations.remove)
        state.push(numId)

        return
      }

      if (change.from === undefined) {
        state.push(MapOperations.add)
        state.push(change.meta.key)
        state.push(numId)
        state.push(change.to)
        return
      }

      state.push(MapOperations.update)
      state.push(numId)
      state.push(change.to)
    })

    return state
  }
}
