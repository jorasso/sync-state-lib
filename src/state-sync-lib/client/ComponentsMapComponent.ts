import { ComponentChangeType, MapOperations } from "../common-types/CommonTypes"
import { Component } from "./Component"

export class ComponentsMapComponent<T extends Component> extends Component {
  map: Map<string, T> = new Map()

  numIdToKey = new Map<number, string>()

  onAddListeners: ((key: string, value: T) => void)[] = []
  onUpdateListeners: ((key: string, value: T) => void)[] = []
  onRemoveListeners: ((key: string) => void)[] = []

  private instanceConstructor: { new (): T }

  constructor(instanceConstructor: { new (): T }) {
    super()
    this.instanceConstructor = instanceConstructor
  }

  private createInstance(): T {
    return new this.instanceConstructor()
  }

  get(key: string) {
    return this.map.get(key)
  }

  has(key: string) {
    return this.map.has(key)
  }

  onAdd(listener: (key: string, value: T) => void) {
    this.onAddListeners.push(listener)
  }

  emitOnAdd(key: string, value: T) {
    for (const listener of this.onAddListeners) {
      listener(key, value)
    }
  }

  onUpdate(listener: (key: string, value: T) => void) {
    this.onUpdateListeners.push(listener)
  }

  emitOnUpdate(key: string, value: T) {
    for (const listener of this.onUpdateListeners) {
      listener(key, value)
    }
  }

  onRemove(listener: (key: string) => void) {
    this.onRemoveListeners.push(listener)
  }

  emitOnRemove(key: string) {
    for (const listener of this.onRemoveListeners) {
      listener(key)
    }
  }

  forEach(cb: (element: T, key: string) => void) {
    this.map.forEach(cb)
  }

  values() {
    return this.map.values()
  }

  size() {
    return this.map.size
  }

  loadCompleteState(state: any[]) {
    this.map.clear()
    this.components.clear()

    for (let i = 2; i < state.length; i += 3) {
      const key = state[i]
      const numId = state[i + 1]
      const data = state[i + 2]

      const component = this.createInstance()

      component.setParent(this, numId, key)
      component.loadCompleteState(data)

      this.components.set(numId, component)

      this.map.set(key, component)

      this.emitOnAdd(key, component)
      component.loadCompleteState(data)
    }
  }

  applyRecentChanges(changes: any[]): void {
    if (changes[0] === ComponentChangeType.complete) {
      this.loadCompleteState(changes)
      return
    }

    if (changes[0] !== ComponentChangeType.partial) {
      return
    }

    for (let i = 1; i < changes.length; i += 1) {
      const operation = changes[i]

      if (operation === MapOperations.add) {
        const key = changes[i + 1]
        const numId = changes[i + 2]
        const data = changes[i + 3]

        const component = this.createInstance()
        component.setParent(this, numId, key)

        // Load complete state first, so the component has proper data before we call onAdd callbacks
        component.loadCompleteState(data)

        this.components.set(numId, component)
        this.map.set(key, component)

        this.emitOnAdd(key, component)

        // Load the component again, so it can fire all the callbacks
        component.loadCompleteState(data)

        i += 3
      } else if (operation === MapOperations.replace) {
        const numId = changes[i + 1]
        const data = changes[i + 2]

        const component = this.components.get(numId)

        component.loadCompleteState(data)

        const key = component.getParentPropertyKey()

        this.emitOnUpdate(key, component as T)

        i += 2
      } else if (operation === MapOperations.remove) {
        const numId = changes[i + 1]

        const component = this.components.get(numId)

        const key = component.getParentPropertyKey()

        this.map.delete(key)
        this.components.delete(numId)

        this.emitOnRemove(key)

        i += 1
      } else if (operation === MapOperations.update) {
        const numId = changes[i + 1]
        const data = changes[i + 2]

        const component = this.components.get(numId)

        component.applyRecentChanges(data)

        const key = component.getParentPropertyKey()

        this.emitOnUpdate(key, component as T)

        i += 1
      }
    }
  }
}
