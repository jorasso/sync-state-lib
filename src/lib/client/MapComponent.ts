import { ComponentChangeType } from "../server/Component"
import { Component } from "./Component"

type AllowedTypes = number | string | boolean

enum MapOperations {
  add = 0,
  remove = 1,
  update = 2,
}

export class MapComponent<T extends AllowedTypes> extends Component {
  map: Map<string, T> = new Map()

  onAddListeners: ((key: string, value: T) => void)[] = []
  onUpdateListeners: ((key: string, value: T) => void)[] = []
  onRemoveListeners: ((key: string) => void)[] = []

  get(key: string) {
    return this.map.get(key)
  }

  has(key: string) {
    return this.map.has(key)
  }

  onAdd(listener: (key: string, value: T) => void) {
    this.onAddListeners.push(listener)
  }

  emitOnAdd(key: string, value: T) {
    for (const listener of this.onAddListeners) {
      listener(key, value)
    }
  }

  onUpdate(listener: (key: string, value: T) => void) {
    this.onUpdateListeners.push(listener)
  }

  emitOnUpdate(key: string, value: T) {
    for (const listener of this.onUpdateListeners) {
      listener(key, value)
    }
  }

  onRemove(listener: (key: string) => void) {
    this.onRemoveListeners.push(listener)
  }

  emitOnRemove(key: string) {
    for (const listener of this.onRemoveListeners) {
      listener(key)
    }
  }

  size() {
    return this.map.size
  }

  loadCompleteState(state: any[]) {
    this.map.clear()
    this.properties.clear()

    for (let i = 2; i < state.length; i += 3) {
      const key = state[i]
      const numId = state[i + 1]
      const value = state[i + 2]

      this.properties.set(numId, key)

      this.map.set(key, value)

      this.emitOnAdd(key, value)
    }
  }

  applyRecentChanges(data: any[]): void {
    if (data[0] === ComponentChangeType.complete) {
      this.loadCompleteState(data)
      return
    }

    if (data[0] !== ComponentChangeType.partial) {
      return
    }

    for (let i = 1; i < data.length; i += 1) {
      const operation = data[i]

      if (operation === MapOperations.add) {
        const key = data[i + 1]
        const numId = data[i + 2]
        const value = data[i + 3]

        this.properties.set(numId, key)
        this.map.set(key, value)

        this.emitOnAdd(key, value)

        i += 3
      } else if (operation === MapOperations.remove) {
        const numId = data[i + 1]
        const key = this.properties.get(numId)

        this.map.delete(key)
        this.properties.delete(numId)

        this.emitOnRemove(key)

        i += 1
      } else if (operation === MapOperations.update) {
        const numId = data[i + 1]
        const value = data[i + 2]

        const key = this.properties.get(numId)

        this.map.set(key, value)

        this.emitOnUpdate(key, value)

        i += 1
      }
    }
  }
}
