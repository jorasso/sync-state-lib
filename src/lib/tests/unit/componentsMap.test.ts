import { expect, test, vi } from "vitest"
import { ComponentsMapComponent as ComponentsMapComponentClient } from "../../client/ComponentsMapComponent"
import { ComponentsMapComponent } from "../../server/ComponentsMapComponent"
import {
  ComplexComponent,
  ComplexComponentClient,
  SimpleComponent,
} from "./TestComponents/TestComponents"

const GetData = () => {
  const client = new ComponentsMapComponentClient<ComplexComponentClient>(
    ComplexComponentClient
  )

  const server = new ComponentsMapComponent<ComplexComponent>(ComplexComponent)

  const componentA = new ComplexComponent()
  componentA.name = "A"
  componentA.subcomponent.name = "subA"
  componentA.subcomponent.id = 1
  componentA.subcomponent.active = true

  server.set("a", componentA)

  const componentB = new ComplexComponent()
  componentB.name = "B"
  componentB.subcomponent.name = "subB"
  componentB.subcomponent.id = 2
  componentB.subcomponent.active = false

  server.set("b", componentB)

  const componentC = new ComplexComponent()
  componentC.name = "C"
  componentC.subcomponent.name = "subC"
  componentC.subcomponent.id = 3
  componentC.subcomponent.active = true

  server.set("c", componentC)

  const state = server.getCompleteState()
  client.loadCompleteState(state)

  server.resetRecentChanges()

  return { server, client, state }
}

test("complete state", () => {
  const { server, client, state } = GetData()

  // Check if the server state is correct
  expect(server.get("a")).toBeDefined()
  expect(server.get("b")).toBeDefined()
  expect(server.get("c")).toBeDefined()

  expect(server.size()).toBe(3)

  expect(server.get("a").name).toBe("A")
  expect(server.get("b").name).toBe("B")
  expect(server.get("c").name).toBe("C")

  expect(server.get("a").subcomponent.name).toBe("subA")
  expect(server.get("b").subcomponent.name).toBe("subB")
  expect(server.get("c").subcomponent.name).toBe("subC")

  expect(server.get("a").subcomponent.id).toBe(1)
  expect(server.get("b").subcomponent.id).toBe(2)
  expect(server.get("c").subcomponent.id).toBe(3)

  expect(server.get("a").subcomponent.active).toBe(true)
  expect(server.get("b").subcomponent.active).toBe(false)
  expect(server.get("c").subcomponent.active).toBe(true)

  // Check if full state data is correct
  expect(state).toMatchSnapshot()

  // Check if the client state is the same as the server state
  expect(client.get("a")).toBeDefined()
  expect(client.get("b")).toBeDefined()
  expect(client.get("c")).toBeDefined()

  expect(client.size()).toBe(3)

  expect(client.get("a").name).toBe(server.get("a").name)
  expect(client.get("b").name).toBe(server.get("b").name)
  expect(client.get("c").name).toBe(server.get("c").name)

  expect(client.get("a").subcomponent.name).toBe(
    server.get("a").subcomponent.name
  )
  expect(client.get("b").subcomponent.name).toBe(
    server.get("b").subcomponent.name
  )
  expect(client.get("c").subcomponent.name).toBe(
    server.get("c").subcomponent.name
  )

  expect(client.get("a").subcomponent.id).toBe(server.get("a").subcomponent.id)
  expect(client.get("b").subcomponent.id).toBe(server.get("b").subcomponent.id)
  expect(client.get("c").subcomponent.id).toBe(server.get("c").subcomponent.id)

  expect(client.get("a").subcomponent.active).toBe(
    server.get("a").subcomponent.active
  )
  expect(client.get("b").subcomponent.active).toBe(
    server.get("b").subcomponent.active
  )
  expect(client.get("c").subcomponent.active).toBe(
    server.get("c").subcomponent.active
  )
})

test("callbacks for initial state", () => {
  const { client, state } = GetData()

  // Create callbacks for the client
  const addCallback = vi.fn()

  const nameUpdateCallback = vi.fn()

  const subcomponentNameUpdateCallback = vi.fn()
  const subcomponentIdUpdateCallback = vi.fn()
  const subcomponentActiveUpdateCallback = vi.fn()

  client.onAdd((key: string, component: ComplexComponentClient) => {
    addCallback(key, component)

    component.on("name", nameUpdateCallback)

    component.subcomponent.on("name", subcomponentNameUpdateCallback)
    component.subcomponent.on("id", subcomponentIdUpdateCallback)
    component.subcomponent.on("active", subcomponentActiveUpdateCallback)
  })

  // Should reset the client state
  client.loadCompleteState(state)

  // Check if the callbacks are called with the correct values
  expect(addCallback).toHaveBeenCalledWith("a", client.get("a"))
  expect(addCallback).toHaveBeenCalledWith("b", client.get("b"))
  expect(addCallback).toHaveBeenCalledWith("c", client.get("c"))

  expect(addCallback).toHaveBeenCalledTimes(3)

  expect(nameUpdateCallback).toHaveBeenCalledTimes(3)
  expect(nameUpdateCallback).toHaveBeenCalledWith("A")
  expect(nameUpdateCallback).toHaveBeenCalledWith("B")
  expect(nameUpdateCallback).toHaveBeenCalledWith("C")

  expect(subcomponentNameUpdateCallback).toHaveBeenCalledTimes(3)
  expect(subcomponentNameUpdateCallback).toHaveBeenCalledWith("subA")
  expect(subcomponentNameUpdateCallback).toHaveBeenCalledWith("subB")
  expect(subcomponentNameUpdateCallback).toHaveBeenCalledWith("subC")

  expect(subcomponentIdUpdateCallback).toHaveBeenCalledTimes(3)
  expect(subcomponentIdUpdateCallback).toHaveBeenCalledWith(1)
  expect(subcomponentIdUpdateCallback).toHaveBeenCalledWith(2)
  expect(subcomponentIdUpdateCallback).toHaveBeenCalledWith(3)
})

test("add component", () => {
  const { server, client } = GetData()

  const addCallback = vi.fn()
  client.onAdd(addCallback)

  const componentD = new ComplexComponent()
  componentD.name = "D"
  componentD.subcomponent.name = "subD"
  componentD.subcomponent.id = 4
  componentD.subcomponent.active = true

  server.set("d", componentD)

  const recentChanges = server.getRecentChanges()
  server.resetRecentChanges()

  expect(recentChanges).toMatchSnapshot()

  client.applyRecentChanges(recentChanges)

  const componentDClient = client.get("d")
  expect(componentDClient).toBeDefined()
  expect(componentDClient?.name).toBe("D")
  expect(componentDClient?.subcomponent.name).toBe("subD")
  expect(componentDClient?.subcomponent.id).toBe(4)
  expect(componentDClient?.subcomponent.active).toBe(true)

  const componentDServer = server.get("d")
  expect(componentDServer).toBeDefined()
  expect(componentDServer?.name).toBe("D")
  expect(componentDServer?.subcomponent.name).toBe("subD")
  expect(componentDServer?.subcomponent.id).toBe(4)
  expect(componentDServer?.subcomponent.active).toBe(true)

  expect(client.size()).toBe(4)
  expect(server.size()).toBe(4)

  expect(addCallback).toBeCalledTimes(1)
  expect(addCallback).toBeCalledWith("d", componentDClient)
})

test("update component", () => {
  const { server, client } = GetData()

  const updateCallback = vi.fn()
  client.onUpdate(updateCallback)

  const componentBClient = client.get("b")

  const nameUpdateCallback = vi.fn()
  componentBClient.on("name", nameUpdateCallback)

  const subcomponentNameUpdateCallback = vi.fn()
  componentBClient.subcomponent.on("name", subcomponentNameUpdateCallback)

  const subcomponentIdUpdateCallback = vi.fn()
  componentBClient.subcomponent.on("id", subcomponentIdUpdateCallback)

  const subcomponentActiveUpdateCallback = vi.fn()
  componentBClient.subcomponent.on("active", subcomponentActiveUpdateCallback)

  const componentBServer = server.get("b")

  componentBServer.name = "B2"
  componentBServer.subcomponent.name = "subB2"
  componentBServer.subcomponent.id = 22
  componentBServer.subcomponent.active = true

  const recentChanges = server.getRecentChanges()
  server.resetRecentChanges()

  expect(recentChanges).toMatchSnapshot()

  client.applyRecentChanges(recentChanges)

  const newComponentBClient = client.get("b")

  expect(newComponentBClient).toBeDefined()

  expect(newComponentBClient.name).toBe("B2")
  expect(newComponentBClient.subcomponent.name).toBe("subB2")
  expect(newComponentBClient.subcomponent.id).toBe(22)
  expect(newComponentBClient.subcomponent.active).toBe(true)

  expect(nameUpdateCallback).toBeCalledTimes(1)
  expect(nameUpdateCallback).toBeCalledWith("B2")

  expect(subcomponentNameUpdateCallback).toBeCalledTimes(1)
  expect(subcomponentNameUpdateCallback).toBeCalledWith("subB2")

  expect(subcomponentIdUpdateCallback).toBeCalledTimes(1)
  expect(subcomponentIdUpdateCallback).toBeCalledWith(22)

  expect(subcomponentActiveUpdateCallback).toBeCalledTimes(1)
  expect(subcomponentActiveUpdateCallback).toBeCalledWith(true)
})

test("delete component", () => {
  const { server, client } = GetData()

  const deleteCallback = vi.fn()
  client.onRemove(deleteCallback)

  server.delete("b")

  const recentChanges = server.getRecentChanges()
  server.resetRecentChanges()

  expect(recentChanges).toMatchSnapshot()

  client.applyRecentChanges(recentChanges)

  expect(client.size()).toBe(2)
  expect(server.size()).toBe(2)

  expect(client.has("a")).toBe(true)
  expect(client.has("b")).toBe(false)
  expect(client.has("c")).toBe(true)

  expect(server.has("a")).toBe(true)
  expect(server.has("b")).toBe(false)
  expect(server.has("c")).toBe(true)

  expect(deleteCallback).toBeCalledTimes(1)
  expect(deleteCallback).toBeCalledWith("b")
})

test("replace component", () => {
  const { server, client } = GetData()

  const updateCallback = vi.fn()
  client.onUpdate(updateCallback)

  const newComponentBServer = new ComplexComponent()
  newComponentBServer.name = "B2"
  newComponentBServer.subcomponent.name = "subB2"
  newComponentBServer.subcomponent.id = 22
  newComponentBServer.subcomponent.active = true

  server.set("b", newComponentBServer)

  const recentChanges = server.getRecentChanges()
  server.resetRecentChanges()

  expect(recentChanges).toMatchSnapshot()

  const componentBClient = client.get("b")

  const nameUpdateCallback = vi.fn()
  componentBClient.on("name", nameUpdateCallback)

  const subcomponentNameUpdateCallback = vi.fn()
  componentBClient.subcomponent.on("name", subcomponentNameUpdateCallback)

  const subcomponentIdUpdateCallback = vi.fn()
  componentBClient.subcomponent.on("id", subcomponentIdUpdateCallback)

  const subcomponentActiveUpdateCallback = vi.fn()
  componentBClient.subcomponent.on("active", subcomponentActiveUpdateCallback)

  client.applyRecentChanges(recentChanges)

  expect(client.size()).toBe(3)
  expect(server.size()).toBe(3)

  expect(client.has("b")).toBe(true)
  expect(server.has("b")).toBe(true)

  expect(client.get("b")).toBe(componentBClient)

  expect(componentBClient).toBeDefined()

  expect(componentBClient.name).toBe("B2")
  expect(componentBClient.subcomponent.name).toBe("subB2")
  expect(componentBClient.subcomponent.id).toBe(22)
  expect(componentBClient.subcomponent.active).toBe(true)

  expect(nameUpdateCallback).toBeCalledTimes(1)
  expect(nameUpdateCallback).toBeCalledWith("B2")

  expect(subcomponentNameUpdateCallback).toBeCalledTimes(1)
  expect(subcomponentNameUpdateCallback).toBeCalledWith("subB2")

  expect(subcomponentIdUpdateCallback).toBeCalledTimes(1)
  expect(subcomponentIdUpdateCallback).toBeCalledWith(22)

  expect(subcomponentActiveUpdateCallback).toBeCalledTimes(1)
  expect(subcomponentActiveUpdateCallback).toBeCalledWith(true)
})

test("reassign subcomponent", () => {
  const { server, client } = GetData()

  const serverSubcomponent = new SimpleComponent()

  const componentBClient = client.get("b")

  server.get("b").subcomponent = serverSubcomponent

  const recentChanges = server.getRecentChanges()
  server.resetRecentChanges()

  const nameUpdateCallback = vi.fn()
  componentBClient.subcomponent.on("name", nameUpdateCallback)

  const idUpdateCallback = vi.fn()
  componentBClient.subcomponent.on("id", idUpdateCallback)

  const activeUpdateCallback = vi.fn()
  componentBClient.subcomponent.on("active", activeUpdateCallback)

  expect(recentChanges).toMatchSnapshot()

  client.applyRecentChanges(recentChanges)

  expect(client.size()).toBe(3)

  expect(componentBClient).toBeDefined()

  expect(componentBClient.subcomponent.name).toBe(serverSubcomponent.name)
  expect(componentBClient.subcomponent.id).toBe(serverSubcomponent.id)
  expect(componentBClient.subcomponent.active).toBe(serverSubcomponent.active)

  expect(nameUpdateCallback).toBeCalledTimes(1)
  expect(nameUpdateCallback).toBeCalledWith(serverSubcomponent.name)

  expect(idUpdateCallback).toBeCalledTimes(1)
  expect(idUpdateCallback).toBeCalledWith(serverSubcomponent.id)

  expect(activeUpdateCallback).toBeCalledTimes(1)
  expect(activeUpdateCallback).toBeCalledWith(serverSubcomponent.active)
})
