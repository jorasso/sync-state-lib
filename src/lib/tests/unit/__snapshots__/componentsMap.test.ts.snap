// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`add component 1`] = `
[
  2,
  0,
  "d",
  3,
  [
    1,
    0,
    "D",
    1,
    [
      1,
      2,
      "subD",
      1,
      4,
      0,
      true,
    ],
  ],
]
`;

exports[`complete state 1`] = `
[
  1,
  3,
  "a",
  0,
  [
    1,
    0,
    "A",
    1,
    [
      1,
      2,
      "subA",
      1,
      1,
      0,
      true,
    ],
  ],
  "b",
  1,
  [
    1,
    0,
    "B",
    1,
    [
      1,
      2,
      "subB",
      1,
      2,
      0,
      false,
    ],
  ],
  "c",
  2,
  [
    1,
    0,
    "C",
    1,
    [
      1,
      2,
      "subC",
      1,
      3,
      0,
      true,
    ],
  ],
]
`;

exports[`delete component 1`] = `
[
  2,
  1,
  1,
]
`;

exports[`reassign subcomponent 1`] = `
[
  2,
  2,
  1,
  [
    2,
    1,
    [
      1,
      2,
      "Some Name",
      1,
      99,
      0,
      true,
    ],
  ],
]
`;

exports[`replace component 1`] = `
[
  2,
  3,
  1,
  [
    1,
    0,
    "B2",
    1,
    [
      1,
      2,
      "subB2",
      1,
      22,
      0,
      true,
    ],
  ],
]
`;

exports[`update component 1`] = `
[
  2,
  2,
  1,
  [
    2,
    0,
    "B2",
    1,
    [
      2,
      2,
      "subB2",
      1,
      22,
      0,
      true,
    ],
  ],
]
`;
